### ISSUE TITLE
Decoding nested struct works in debug (default) build, fails in release build

### ISSUE THREAD

**Issue by kookman** [andrejp88/toml-foolery#1]
I'm having a problem decoding the following TOML on ARM (32 bit).
``` toml
[[tanks]]
mqttTopic = "sensors/MusicRoomTank/depthCm"
tankDepthCm = 48
microsToCmFactor = 55.0

[[tanks]]
mqttTopic = "sensors/MainTank/depthCm"
tankDepthCm = 200
microsToCmFactor = 55.0
```

The code (reduced) to decode:
``` d
/+ dub.sdl:
    name "tomltest"
    dependency "toml-foolery" version="~>1.0"
+/

import toml_foolery;
import std.stdio;
import std.file;

struct Config
{
    TankSensor[] tanks;

    struct TankSensor
    {
        string mqttTopic;
        int tankDepthCm;
        float microsToCmFactor;
        int countMeasures;
        int runningTotal;
    }
}


void main()
{
    auto config = readText("tankconfig.toml").parseToml!Config;
    writefln("Tanks config length: %s, content: %s", config.tanks.length, config.tanks);
}
```

The same code/config works fine on x86_64.

**Comment by kookman** [comment]
Actually, this might be a cross-compilation issue. Seems to work compiling on Raspberry Pi using LDC 1.17.

**Comment by kookman** [comment]
No. The cross-compilation theory is a red herring. Looking further into it, the problem appears with a release build, but not with a debug build:
```
pi@pidrie:~/code/anvil/tankdepth $ dub build --single test.d
Performing "debug" build using /home/pi/Downloads/ldc2-1.17.0-linux-armhf/bin/ldc2 for arm, arm_hardfloat.
datefmt 1.0.4: target for configuration "library" is up to date.
pegged 0.4.5: target for configuration "default" is up to date.
toml-foolery 1.0.0: target for configuration "library" is up to date.
tomltest ~master: building configuration "application"...
To force a rebuild of up-to-date targets, run again with --force.
pi@pidrie:~/code/anvil/tankdepth $ ./tomltest
Tanks config length: 2, content: [TankSensor("sensors/MusicRoomTank/depthCm", 48, 55, 0, 0), TankSensor("sensors/MainTank/depthCm", 200, 55, 0, 0)]
pi@pidrie:~/code/anvil/tankdepth $ dub build --single --build=release test.d
Performing "release" build using /home/pi/Downloads/ldc2-1.17.0-linux-armhf/bin/ldc2 for arm, arm_hardfloat.
datefmt 1.0.4: target for configuration "library" is up to date.
pegged 0.4.5: target for configuration "default" is up to date.
toml-foolery 1.0.0: target for configuration "library" is up to date.
tomltest ~master: building configuration "application"...
To force a rebuild of up-to-date targets, run again with --force.
pi@pidrie:~/code/anvil/tankdepth $ ./tomltest
toml_foolery.decode.exceptions.TomlDecodingException@/home/pi/.dub/packages/toml-foolery-1.0.0/toml-foolery/source/toml_foolery/decode/set_data.d(114): Could not place value "sensors/MusicRoomTank/depthCm" of type "string" inside field "tanks" of type "TankSensor[]" in struct "Config". This might be a bug — please file a report.
----------------
??:? [0x1817f7]
??:? [0x18ce93]
??:? [0x1719cf]
??:? [0x10eaf]
??:? [0x1055b]
??:? [0x10017]
??:? [0xf93f]
??:? [0xbbbb]
??:? [0x171603]
??:? [0x1714bb]
??:? [0x17134b]
??:? __libc_start_main [0x76d52717]
pi@pidrie:~/code/anvil/tankdepth $
```
Note `test.d` contains the code in my comment above, and `tankconfig.toml` contains the TOML config shown above.

**Comment by kookman** [comment]
Just to confirm, platform x86_64 and LDC latest version (1.27-beta) gives exactly the same result: Works in debug build, fails in release.

```
dev@devbox:~/code/gitlab/anvil/tankdepth$ dub build --single test.d
Performing "debug" build using /home/dev/Downloads/ldc2-1.27.0-beta1-linux-x86_64/bin/ldc2 for x86_64.
datefmt 1.0.4: target for configuration "library" is up to date.
pegged 0.4.5: target for configuration "default" is up to date.
toml-foolery 1.0.0: target for configuration "library" is up to date.
tomltest ~master: building configuration "application"...
Linking...
To force a rebuild of up-to-date targets, run again with --force.
dev@devbox:~/code/gitlab/anvil/tankdepth$ ./tomltest
Tanks config length: 2, content: [TankSensor("sensors/MusicRoomTank/depthCm", 48, 55, 0, 0), TankSensor("sensors/MainTank/depthCm", 200, 55, 0, 0)]
dev@devbox:~/code/gitlab/anvil/tankdepth$ dub build --single --build=release test.d
Performing "release" build using /home/dev/Downloads/ldc2-1.27.0-beta1-linux-x86_64/bin/ldc2 for x86_64.
datefmt 1.0.4: building configuration "library"...
/home/dev/.dub/packages/datefmt-1.0.4/datefmt/source/datefmt.d(410,17): Deprecation: variable `i` is shadowing variable `datefmt.Interpreter.parse.__foreachbody4.i`. Rename the `foreach` variable.
pegged 0.4.5: building configuration "default"...
toml-foolery 1.0.0: building configuration "library"...
tomltest ~master: building configuration "application"...
Linking...
dev@devbox:~/code/gitlab/anvil/tankdepth$ ./tomltest
toml_foolery.decode.exceptions.TomlDecodingException@/home/dev/.dub/packages/toml-foolery-1.0.0/toml-foolery/source/toml_foolery/decode/set_data.d(114): Could not place value "sensors/MusicRoomTank/depthCm" of type "string" inside field "tanks" of type "TankSensor[]" in struct "Config". This might be a bug — please file a report.
----------------
??:? [0x560d9e0b2525]
??:? [0x560d9e0b3fa6]
??:? [0x560d9e09720f]
??:? [0x560d9deec4e0]
??:? [0x560d9deeaf81]
??:? [0x560d9dee499b]
??:? [0x560d9e096ef7]
??:? [0x560d9e096df4]
??:? [0x560d9e096c4d]
??:? __libc_start_main [0x7f0d4978009a]
??:? [0x560d9dee48a9]
dev@devbox:~/code/gitlab/anvil/tankdepth$
```

**Comment by andrej-petrovic** [comment]
Hmm, strange. Thanks for the detailed report. I'll have some time during the weekend to dig into it.

**Comment by andrej-petrovic** [comment]
I pushed a fix, it's available in v1.0.1. The example you gave now works in both debug and release modes. Specifically, the problem showed up when assertions were disabled. Let me know if you still have trouble :)

**Comment by kookman** [comment]
Verified this works with v1.0.1. Thanks @andrejp88!
